File: ts-langchain/src/tools/web-extract.ts
/**
 * web.extract — 网页结构化信息提取
 *
 * 在 web.read 的 SSRF 防护基础上抓取网页，并从重复卡片、表格和字段标签中提取结构化数据。
 */

import { DynamicStructuredTool } from "@langchain/core/tools";
import { z } from "zod";

import type { ToolDescriptor } from "./contracts.js";
import { isSafeUrl, validateNetworkUrl } from "./web-read.js";

const MAX_RESPONSE_CHARS = 200 * 1024;

export const webExtractInputSchema = z.object({
  url: z.string().url().describe("要提取数据的网页 URL"),
  fields: z.array(z.string().min(1).max(80)).min(1).max(20).describe("需要提取的字段名，如 name、price、rating"),
});

export const webExtractDescriptor: ToolDescriptor = {
  name: "web.extract",
  version: "1.0.0",
  title: "网页结构化提取",
  description: "从指定网页提取结构化字段，例如产品名称、价格和评分。适合批量列表页，不用于文章总结。",
  category: "READ",
  risk_level: "R1",
  side_effect: "read",
  timeout_ms: 15000,
  required_permissions: ["web.extract"],
  data_classification: ["public"],
  owner: "tools",
  tags: ["web", "extract", "structured"],
  input_schema: webExtractInputSchema,
};

// 解码网页中常见的 HTML 实体。
function decodeHtml(text: string): string {
  return text
    .replace(/&nbsp;/gi, " ")
    .replace(/&amp;/gi, "&")
    .replace(/&lt;/gi, "<")
    .replace(/&gt;/gi, ">")
    .replace(/&quot;/gi, '"')
    .replace(/&#39;/gi, "'")
    .replace(/&#(\d+);/g, (_, code: string) => String.fromCharCode(Number(code)));
}

// 清理 HTML 标签、脚本和样式，保留可供字段匹配的文本。
function cleanHtml(html: string): string {
  return decodeHtml(
    html
      .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, " ")
      .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, " ")
      .replace(/<[^>]+>/g, " ")
      .replace(/[ \t\r\n]+/g, " ")
      .trim(),
  );
}

// 生成字段的宽松匹配键，兼容 class、id 和 data-field 命名。
function fieldKeys(field: string): string[] {
  const normalized = field.trim().toLowerCase().replace(/[_-]+/g, " ");
  const aliases: Record<string, string[]> = {
    name: ["name", "title", "product name"],
    title: ["title", "name"],
    price: ["price", "cost", "amount"],
    rating: ["rating", "score", "stars"],
  };
  return [normalized, ...(aliases[normalized] ?? [])];
}

// 从一个 HTML 片段中按字段名提取值。
function extractField(fragment: string, field: string): string {
  const keys = fieldKeys(field).map((key) => key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
  const keyPattern = keys.join("|");
  const elementPattern = new RegExp(
    `<([a-z0-9]+)[^>]*(?:class|id|data-field|itemprop)=["'][^"']*(?:${keyPattern})[^"']*["'][^>]*>([\\s\\S]*?)<\\/\\1>`,
    "i",
  );
  const elementMatch = fragment.match(elementPattern);
  if (elementMatch) return cleanHtml(elementMatch[2]).slice(0, 500);

  const labelPattern = new RegExp(`(?:${keyPattern})\\s*[:：]\\s*([^|;,]+)`, "i");
  const labelMatch = cleanHtml(fragment).match(labelPattern);
  return labelMatch?.[1]?.trim().slice(0, 500) ?? "";
}

// 提取表格中的行，并以表头作为字段名。
function extractTableItems(html: string, fields: string[]): Array<Record<string, string>> {
  const tables: Array<Record<string, string>> = [];
  const tableMatches = html.matchAll(/<table[^>]*>([\s\S]*?)<\/table>/gi);
  for (const tableMatch of tableMatches) {
    const rows = [...tableMatch[1].matchAll(/<tr[^>]*>([\s\S]*?)<\/tr>/gi)].map((row) =>
      [...row[1].matchAll(/<t[dh][^>]*>([\s\S]*?)<\/t[dh]>/gi)].map((cell) => cleanHtml(cell[1])),
    );
    if (rows.length < 2) continue;
    const headers = rows[0].map((header) => header.toLowerCase());
    for (const row of rows.slice(1)) {
      const item: Record<string, string> = {};
      for (const field of fields) {
        const index = headers.findIndex((header) => fieldKeys(field).some((key) => header.includes(key)));
        if (index >= 0 && row[index]) item[field] = row[index].slice(0, 500);
      }
      if (Object.keys(item).length > 0) tables.push(item);
    }
  }
  return tables;
}

// 提取产品卡片、列表项或语义化 article 中的字段。
function extractBlockItems(html: string, fields: string[]): Array<Record<string, string>> {
  const items: Array<Record<string, string>> = [];
  const blockPattern = /<(article|li|div|section)[^>]*(?:class|id|itemtype)=["'][^"']*(?:product|item|card|listing|offer)[^"']*["'][^>]*>([\s\S]*?)<\/\1>/gi;
  for (const match of html.matchAll(blockPattern)) {
    const item: Record<string, string> = {};
    for (const field of fields) {
      const value = extractField(match[0], field);
      if (value) item[field] = value;
    }
    if (Object.keys(item).length > 0) items.push(item);
  }
  return items;
}

// 从网页 HTML 中提取请求字段并生成结构化结果。
export function extractStructuredItems(html: string, fields: string[]): Array<Record<string, string>> {
  const tableItems = extractTableItems(html, fields);
  if (tableItems.length > 0) return tableItems;
  const blockItems = extractBlockItems(html, fields);
  if (blockItems.length > 0) return blockItems;

  const fallback: Record<string, string> = {};
  for (const field of fields) {
    const value = extractField(html, field);
    if (value) fallback[field] = value;
  }
  return Object.keys(fallback).length > 0 ? [fallback] : [];
}

// 在每次重定向前执行 SSRF 安全检查并读取网页正文。
async function fetchSafeHtml(url: string): Promise<string> {
  const initialCheck = await validateNetworkUrl(url);
  if (!initialCheck.safe) throw new Error(`URL 安全拒绝：${initialCheck.reason}`);

  let currentUrl = url;
  for (let redirects = 0; redirects <= 3; redirects++) {
    const targetCheck = await validateNetworkUrl(currentUrl);
    if (!targetCheck.safe) throw new Error(`重定向目标 URL 安全拒绝：${targetCheck.reason}`);
    const response = await fetch(currentUrl, {
      headers: { "User-Agent": "Mozilla/5.0 (compatible; AI-Agent/1.0)", Accept: "text/html,text/plain,application/json,*/*" },
      redirect: "manual",
    });
    if ([301, 302, 303, 307, 308].includes(response.status)) {
      const location = response.headers.get("location");
      await response.body?.cancel();
      if (!location) throw new Error("重定向响应缺少 Location");
      if (redirects === 3) throw new Error("重定向次数超过 3 次");
      currentUrl = new URL(location, currentUrl).toString();
      continue;
    }
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    const contentType = (response.headers.get("content-type") || "").split(";", 1)[0].toLowerCase();
    if (contentType && !contentType.startsWith("text/") && !["application/json", "application/xhtml+xml"].includes(contentType)) {
      throw new Error(`不支持的内容类型：${contentType}`);
    }
    return (await response.text()).slice(0, MAX_RESPONSE_CHARS);
  }
  throw new Error("重定向次数超过 3 次");
}

export const webExtractTool: DynamicStructuredTool = new DynamicStructuredTool({
  name: "web_extract",
  description: "从网页列表或表格中提取结构化字段，例如产品名称、价格和评分。",
  schema: webExtractInputSchema,
  func: async ({ url, fields }) => {
    if (!isSafeUrl(url).safe) {
      return JSON.stringify({ items: [], error: "URL 安全拒绝" });
    }
    try {
      const html = await fetchSafeHtml(url);
      return JSON.stringify({ items: extractStructuredItems(html, fields) });
    } catch (error) {
      return JSON.stringify({
        items: [],
        error: `网页结构化提取失败：${error instanceof Error ? error.message : "未知错误"}`,
      });
    }
  },
});

File: py-langgraph/src/tools/web_extract.py
"""web.extract — 从网页中提取结构化字段。"""

from __future__ import annotations

import html as html_module
import json
import re

import httpx
from langchain_core.tools import tool
from pydantic import BaseModel, Field

from src.tools.contracts import ToolDescriptor
from src.tools.web_read import _is_safe_url

MAX_RESPONSE_CHARS = 200 * 1024


class WebExtractInput(BaseModel):
    """网页结构化提取工具的输入模型。"""

    url: str = Field(description="要提取数据的网页 URL")
    fields: list[str] = Field(min_length=1, max_length=20, description="需要提取的字段名")


_DESCRIPTOR = ToolDescriptor(
    name="web.extract",
    version="1.0.0",
    title="网页结构化提取",
    description="从指定网页提取结构化字段，例如产品名称、价格和评分。适合批量列表页，不用于文章总结。",
    category="READ",
    risk_level="R1",
    side_effect="read",
    timeout_ms=15000,
    required_permissions=["web.extract"],
    data_classification=["public"],
    owner="tools",
    tags=["web", "extract", "structured"],
)


# 解码网页中常见的 HTML 实体。
def _decode_html(value: str) -> str:
    return html_module.unescape(value)


# 清理 HTML 标签、脚本和样式，保留可匹配的文本。
def _clean_html(value: str) -> str:
    text = re.sub(r"<script[^>]*>[\s\S]*?</script>", " ", value, flags=re.IGNORECASE)
    text = re.sub(r"<style[^>]*>[\s\S]*?</style>", " ", text, flags=re.IGNORECASE)
    text = re.sub(r"<[^>]+>", " ", text)
    return re.sub(r"\s+", " ", _decode_html(text)).strip()


# 生成字段的宽松匹配键，兼容常见产品字段别名。
def _field_keys(field: str) -> list[str]:
    normalized = field.strip().lower().replace("_", " ").replace("-", " ")
    aliases = {
        "name": ["name", "title", "product name"],
        "title": ["title", "name"],
        "price": ["price", "cost", "amount"],
        "rating": ["rating", "score", "stars"],
    }
    return [normalized, *aliases.get(normalized, [])]


# 从一个 HTML 片段中按字段名提取值。
def _extract_field(fragment: str, field: str) -> str:
    keys = [re.escape(key) for key in _field_keys(field)]
    key_pattern = "|".join(keys)
    element_pattern = re.compile(
        rf'<([a-z0-9]+)[^>]*(?:class|id|data-field|itemprop)=["\'][^"\']*(?:{key_pattern})[^"\']*["\'][^>]*>([\s\S]*?)</\1>',
        re.IGNORECASE,
    )
    match = element_pattern.search(fragment)
    if match:
        return _clean_html(match.group(2))[:500]
    label_match = re.search(rf"(?:{key_pattern})\s*[:：]\s*([^|;,]+)", _clean_html(fragment), re.IGNORECASE)
    return label_match.group(1).strip()[:500] if label_match else ""


# 提取 HTML 表格中的数据行。
def _extract_table_items(page: str, fields: list[str]) -> list[dict[str, str]]:
    items: list[dict[str, str]] = []
    for table_match in re.finditer(r"<table[^>]*>([\s\S]*?)</table>", page, re.IGNORECASE):
        rows = []
        for row_match in re.finditer(r"<tr[^>]*>([\s\S]*?)</tr>", table_match.group(1), re.IGNORECASE):
            rows.append([
                _clean_html(cell.group(1))
                for cell in re.finditer(r"<t[dh][^>]*>([\s\S]*?)</t[dh]>", row_match.group(1), re.IGNORECASE)
            ])
        if len(rows) < 2:
            continue
        headers = [header.lower() for header in rows[0]]
        for row in rows[1:]:
            item: dict[str, str] = {}
            for field in fields:
                index = next((i for i, header in enumerate(headers) if any(key in header for key in _field_keys(field))), -1)
                if index >= 0 and index < len(row) and row[index]:
                    item[field] = row[index][:500]
            if item:
                items.append(item)
    return items


# 提取产品卡片、列表项或 article 中的字段。
def _extract_block_items(page: str, fields: list[str]) -> list[dict[str, str]]:
    items: list[dict[str, str]] = []
    pattern = re.compile(
        r'<(article|li|div|section)[^>]*(?:class|id|itemtype)=["\'][^"\']*(?:product|item|card|listing|offer)[^"\']*["\'][^>]*>([\s\S]*?)</\1>',
        re.IGNORECASE,
    )
    for match in pattern.finditer(page):
        item = {}
        for field in fields:
            value = _extract_field(match.group(0), field)
            if value:
                item[field] = value
        if item:
            items.append(item)
    return items


# 从网页 HTML 中提取请求字段。
def extract_structured_items(page: str, fields: list[str]) -> list[dict[str, str]]:
    """从网页 HTML 中提取结构化字段，供测试和工具复用。"""
    table_items = _extract_table_items(page, fields)
    if table_items:
        return table_items
    block_items = _extract_block_items(page, fields)
    if block_items:
        return block_items
    fallback = {field: value for field in fields if (value := _extract_field(page, field))}
    return [fallback] if fallback else []


# 安全抓取网页正文并复检每个重定向目标。
def _fetch_safe_html(url: str) -> str:
    safe, reason = _is_safe_url(url)
    if not safe:
        raise ValueError(f"URL 安全拒绝：{reason}")
    current_url = url
    with httpx.Client(timeout=10, follow_redirects=False) as client:
        for redirects in range(4):
            safe, reason = _is_safe_url(current_url)
            if not safe:
                raise ValueError(f"重定向目标 URL 安全拒绝：{reason}")
            response = client.get(
                current_url,
                headers={"User-Agent": "Mozilla/5.0 (compatible; AI-Agent/1.0)", "Accept": "text/html,text/plain,application/json,*/*"},
            )
            if response.status_code in {301, 302, 303, 307, 308}:
                location = response.headers.get("location")
                if not location:
                    raise ValueError("重定向响应缺少 Location")
                current_url = str(httpx.URL(current_url).join(location))
                continue
            response.raise_for_status()
            content_type = response.headers.get("content-type", "").split(";", 1)[0].lower()
            if content_type and not (content_type.startswith("text/") or content_type in {"application/json", "application/xhtml+xml"}):
                raise ValueError(f"不支持的内容类型：{content_type}")
            return response.text[:MAX_RESPONSE_CHARS]
    raise ValueError("重定向次数超过 3 次")


# 从指定网页提取结构化字段。
@tool(args_schema=WebExtractInput)
# 执行 web extract 对应的业务逻辑
def web_extract(url: str, fields: list[str]) -> str:
    """从网页列表或表格中提取结构化字段。"""
    try:
        return json.dumps({"items": extract_structured_items(_fetch_safe_html(url), fields)}, ensure_ascii=False)
    except Exception as exc:
        return json.dumps({"items": [], "error": f"网页结构化提取失败：{exc!s}"}, ensure_ascii=False)


__all__ = ["_DESCRIPTOR", "WebExtractInput", "extract_structured_items", "web_extract"]
